### ISSUE TITLE
A question about the discrepancy between paper’s formula and code implementation

### ISSUE THREAD

**Issue by SufeyifanWu** [patrickpynadath1/dab#1]
Hi! Thank you so much for sharing your wonderful work and open-sourcing the code!

I do have a question about a seeming discrepancy between the paper’s theoretical formulation and the actual implementation:

1. **Paper’s Formulation**  
   In the paper, you propose a target distribution of the form:  
   ![公式](https://latex.codecogs.com/png.latex?P(Y,B%20%7C%20X)%20%5Cpropto%20P%5E%7BLM%7D(Y%20%7C%20X%2CB)%5C%2C%5Cexp%5Cbigl(f(B%20%7C%20X)%5Cbigr))  
   indicating the constraint or energy function ![公式](https://latex.codecogs.com/png.latex?f) is a function of ![公式](https://latex.codecogs.com/png.latex?B) (and ![公式](https://latex.codecogs.com/png.latex?X)), rather than the final text ![公式](https://latex.codecogs.com/png.latex?Y). This suggests that the scoring mechanism focuses on ![公式](https://latex.codecogs.com/png.latex?B).

2. **Code Implementation**  
   Upon running the code, however, it appears the actual loss/gradient is computed based on the final text ![公式](https://latex.codecogs.com/png.latex?Y). 
```
        onehot_generates = onehot_generates.detach()
        onehot_generates.requires_grad = True

        dis_embs = torch.matmul(
            onehot_generates, self.discriminator.get_input_embeddings().weight
        )
        senti_logits = self.discriminator(
            inputs_embeds=dis_embs, labels=self.labels.repeat(dis_embs.shape[0])
        ).logits
        if self.is_detoxic:
            senti_losses = -(senti_logits[:, 0] - senti_logits[:, 1])
        else:
            if self.attribute == 'pos': 
                senti_losses = -(senti_logits[:, 1] - senti_logits[:, 0])
            else: 
                senti_losses = -(senti_logits[:, 0] - senti_logits[:, 1])
        senti_loss = torch.sum(senti_losses)
        
        loss = senti_loss 
        # print("senti_loss:", senti_loss)
        # import pdb; pdb.set_trace()

        return (loss, output_ids, onehot_generates, gpt_logit, senti_losses)

```
The onehot_generates seem come from cur_sampled_next_token  ,which appears to represent the generated  Y
```
                cur_sampled_next_token = (
                    cur_sampled_next_token_onehot
                    - cur_sampled_next_token_soft.detach()
                    + cur_sampled_next_token_soft
                )            
sampled_next_tokens = torch.cat(
                (sampled_next_tokens, cur_sampled_next_token.unsqueeze(1)), dim=1
            )
```
Could you clarify the following?

- How does the formulation ![公式](https://latex.codecogs.com/png.latex?%5Cexp(f(B%20%7C%20X))) in the paper align with a model that seemingly scores the final generated text ![公式](https://latex.codecogs.com/png.latex?Y) in the code?  
- Is the actual scoring function more like ![公式](https://latex.codecogs.com/png.latex?f(Y%20%7C%20B%2C%20X)) in practice?  

I’d really appreciate any clarification on these points. Thank you again for your fantastic work!

**Comment by patrickpynadath1** [comment]
Hi! Thanks for your interest. It means a lot. 

The reasoning for this is a bit nuanced. The first equation you shared -- represents the joint distribution. We apply a form of Gibbs sampling to this target distribution. We have the following conditional distribution for P(B|X, Y). 

<img width="553" alt="Image" src="https://github.com/user-attachments/assets/85e08eb2-c083-400f-8f7b-5c4b2f2acadd" />

This distribution involves two terms -- one that places mass on B that increase the likelihood of the give Y, and another that places mass based on the constraint value. 

In practice, this distribution is intractable -- it would require us to compute P(Y|B, X) for all possible B. We get around this by noting that the P^{LM}(Y|B, X) term places higher mass on bias sequences B that are likely to result in the given Y. An easy way to satisfy this is to initialize B=Y, since the bias sequence that maximizes the probability of generating Y via biased auto-regressive generation is Y itself. As a result, the sampling problem becomes a bit easier, since we only need to worry about increasing / moving towards sequences with good f(B|X), since we initialize close to samples that already have good P(Y|B, X). Thus in the implementation, when computing the gradient, we use the generated sequence Y. We explain this logic in section 4.1, under the paragraph "P(B|X, Y)", and in Figure 2, where we show that the previous response sequence Y is what is used when sampling the next bias sequence B. 

Thanks again for your interest, and please continue to reach out if you have additional questions!

**Comment by SufeyifanWu** [comment]
Hello again! Thank you for the clarifications so far.

I do understand that initializing `B = Y` can be viewed as a reasonable approximation to satisfy `P(Y | X, B)`, although I wonder if it might be more accurate to use something like a Metropolis–Hastings step to correct the sampled `B` (e.g., using an acceptance ratio of the form:
![公式](https://latex.codecogs.com/png.latex?a%20%3D%20%5Cmin%5Cleft%28%201%2C%20%5Cfrac%7BP%5E%7BLM%7D%28Y%20%7C%20X%2C%20B_%7Bprop%7D%29%20f%28B_%7Bprop%7D%20%7C%20X%29%7D%7BP%5E%7BLM%7D%28Y%20%7C%20X%2C%20B_%7Bold%7D%29%20f%28B_%7Bold%7D%20%7C%20X%29%7D%5Cright%29 ))

However, my main confusion is not about `P(Y | X, B)` but rather about `f(B | X)`.

Even if we start with `B = Y`, once we begin sampling or updating `B`, it may diverge from `Y`. Yet, in the code, the constraint term `f(B | X)` seems to be derived from a discriminator or classifier applied to the final generated text `Y'`, rather than directly to `B`. Could you clarify how, in practice, you reconcile this with the theoretical formulation `f(B | X)`? Does it mean that we believe `f(B|X)` \approx `f(Y'|X)`, where `Y'` is generated from B and original Y?

I’d greatly appreciate any extra detail on how this part of the implementation maps back to the theoretical framework. Thank you again, and I look forward to your thoughts!

**Comment by SufeyifanWu** [comment]
And another quick question regarding the controlled text generation process.

While running sentiment-controlled experiments, I noticed that during each step of sampling, the controlled generation of the final ` Y' ` is not very stable —some samples within the batch violate the control constraint. 

From analyzing the code, I found that instead of directly using the sampled ` Y' `, the implementation selects the one with the lowest discriminator loss among multiple samples for each batch position. That is, instead of simply taking the generated` Y' `, the model performs multiple sampling passes and chooses the one with the smallest discriminator loss for each batch index.

This approach seems to introduce some inefficiencies, and I see two potential issues with it:

1. The distribution of` B ` does not become more stable as the number of samples increases.  
   - Ideally, repeated sampling should refine \( B \), making it converge to a more stable distribution. However, some samples within the batch still violate the control constraint, even in the final sampling step

2. Using the discriminator to directly correct the sampled results might be problematic.
   - The correction step seems to ignore the fluency constraint imposed by `P(Y \mid X, B)`.  
   - For example, if one sample has a loss of `-10` while another has a loss of `-9.9`, the model will always choose the `-10` sample, even though the`-9.9` sample might be **more fluent** in terms of language quality.  
   - You demonstrated that the PPL is lower than other methods, which is really impressive. However, this evaluation is done at the individual sentence level. While analyzing the generated outputs in a batch, I noticed that many sentences exhibit excessively high repetition within the batch.  
  
So theoretically, incorporating an appropriate proposal distribution after sampling B to perform an accept-reject step might help stabilize the sampled B. I’m curious—why did you ultimately choose to implement the approach where the selection is based on a greedy, locally optimal loss instead? Would love to hear your thoughts on this!

**Comment by patrickpynadath1** [comment]
Hello, 

### Confusion about Constraint
In response to your first point, it may be helpful to view the psuedo-code we include in the appendix: 

![Image](https://github.com/user-attachments/assets/78d262f2-2d2d-4b1f-8abe-99064ecfe259)

There are two primary steps in our Gibbs sampling algorithm: sampling P(Y|B, X); and then sampling P(B|Y,X). Sampling P(Y|B,X) is fine as we can do this via auto-regressive generation. However, when sampling P(B|Y,X), we encounter an issue: this distribution depends not only on the constraint function, but also P(Y|B, X). The application of the initialization trick allows us to ''ignore'' this component of the distribution and sample only based off of the constraint. Once we obtain a new bias sequence, we sample a new Y by applying biased auto-regressive generation, and repeat for some number of steps.

In your response, you mention that you are not concerned with P(Y|B, X) and are confused about f(B|X). However, it is important to note that P(Y|B, X) shows up in both steps of the Gibbs sampling algorithm: first, we sample directly from P(Y|B,X) to obtain Y; and when sampling the bias sequence B, we still need to consider P(Y|B,X). So even though we are sampling B in order to improve the constraint, as indicated in our derivation in section 4, the distribution over B places more mass on bias sequences that increase the likelihood of the conditioned Y. 

In conclusion, when sampling the bias sequence B, our goal is NOT to sample from f(B|X), as that is NOT the true conditional distribution of B given the previous response sequence Y. If we did this, we would NOT be sampling from the joint distribution as defined in equation 5. When producing a bias sequence, our goal is to sample from f(B|X) P(Y|B, X). 

It would be difficult to directly sample this, as it would require another auto-regressive step for every possible bias sequence B. This would be bad, as we already do an autoregressive step when sampling Y. To avoid this, we initialize B=Y and then sample B from f(B=Y|X). So each sampling step of the algorithm can be described as sampling Y from P(Y|B,X), sampling B from P(B|Y, X) \propto f(B|X) P(Y|B,X) by initializing B=Y and sampling from f(B=Y|X). 

### Greedy Selection
In regards to your first point, this instability arises from the difference between optimization and sampling: while optimization is only concerned with getting better samples (more likely), sampling is more concerned with the overall accuracy in terms of the entire target distribution. This necessarily includes low probability samples -- if our sampler completely ignores these regions and never generates a single low probability sample, then our approximation of the target distribution attaches zero mass to these samples. This is obviously inaccurate, as these samples have non-zero mass under the target distribution. Furthermore, there are far more samples in the total support that violate the constraint as opposed to samples that satisfy the constraint. Thus in applying a sampling algorithm, we will see low probability samples occur. We still wish to use a sampling algorithm as these have improved exploration capabilities when compared to optimization algorithms -- see figure 3 in section 4.3 for the strengths of this approach in terms of average unique tokens discovered and average hops per sampling step. We compare to BOLT, which uses the ADAM optimizer to tune continuous bias sequences, which makes it resemble an optimization algorithm. 

Could you also clarify by what you mean "stable"? Do you mean that the batch average tends to blow up over sampling periods; or do you mean that there is a high variance among the batch throughout sampling? The first behavior might be a bug, in which case I will take a look at it. The second behavior is actually fine -- we can attribute this to the exploration of the sampler. This is why the greedy selection is implemented. 

In regards to the greedy selection, you are correct. We do select the samples that have the best constraint value, which does not explicitly incorporate the fluency constraint. However, each sample is produced via biased auto-regressive generation, which implicitly enforces the fluency constraint for each generation step. This is supported by empirical results in section 4.3, where we show that the fluency is stable over a long duration of sampling steps. 

### MH Acceptance Step 
I am not sure the MH step you have given is correct -- it seems to be excluding the ratio of the forwards and backwards proposal distributions. It would should look something like below: 

![Image](https://github.com/user-attachments/assets/2b6dc727-6c8f-4b6f-9307-298d11639c66)

Here, I denote Y, B as the original response sequence and bias sequence; Y', B' as the newly sampled response sequence and bias sequence; and q as the transition probabilities. The metropolis hastings correction you wrote assumes that the forward and backwards transition probabilities are equal, which is only true if we were using a uniform proposal distribution. 

The issue with using the MH step is a result of the ratio of the transition probabilities, as this tends to be very low for highly asymmetric proposal functions. 

It might be a good idea to modify the final selection criteria to include the fluency constraint -- I just did not do this since the results were fairly good without adding this additional step. However, it could be useful in practice. 

If you have further questions related to the algorithm and why it is implemented in this way, I encourage you to reach out to me through ppynadat@purdue.edu. It may also be helpful to have a quick zoom call so I could clarify any concerns or doubts you have quicker.

**Comment by SufeyifanWu** [comment]
Thank you so much again for your incredibly thoughtful and thorough reply! I truly appreciate the time you took to walk through everything in such detail.

I’ll take some more time to carefully think through the points you raised as there are quite a few nuanced and tricky parts here.

Wishing you all the best with your ongoing research and work!
